Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/UpgradesController.cs

```diff
@@ -0,0 +1,228 @@
+
+            return this.Request.CreateResponse(HttpStatusCode.NoContent);
+        }
+
+        /// <summary>Starts a local upgrade.</summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>An empty success response.</returns>
+        [HttpPost]
+        public async Task<HttpResponseMessage> StartFirstUpgrade(CancellationToken cancellationToken)
+        {
+            var upgrades = await this.localUpgradeService.GetLocalUpgrades(cancellationToken);
+            if (!upgrades.Any(u => u.IsValid && !u.IsOutdated))
+            {
+                return this.Request.CreateResponse(HttpStatusCode.BadRequest, new { message = "There are no upgrades to apply", });
+            }
+
+            await this.localUpgradeService.StartLocalUpgrade(upgrades, cancellationToken);
+
+            return this.Request.CreateResponse(HttpStatusCode.NoContent);
+        }
+
+        /// <summary>Deletes a local upgrade package.</summary>
+        /// <param name="data">The DTO containing the packageName.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>An empty success response.</returns>
+        [HttpPost]
+        public async Task<HttpResponseMessage> Delete(UpgradePackageRequestDto data, CancellationToken cancellationToken)
+        {
+            var upgrades = await this.localUpgradeService.GetLocalUpgrades(cancellationToken);
+            var upgrade = upgrades.FirstOrDefault(u => u.PackageName.Equals(data.PackageName, StringComparison.InvariantCultureIgnoreCase));
+            try
+            {
+                var packagePath = Path.Combine(this.applicationStatusInfo.ApplicationMapPath, "App_Data", "Upgrade", upgrade.PackageName + ".zip");
+                if (File.Exists(packagePath))
+                {
+                    File.Delete(packagePath);
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex);
+                return this.Request.CreateResponse(HttpStatusCode.InternalServerError, new { message = ex.Message, });
+            }
+
+            return this.Request.CreateResponse(HttpStatusCode.NoContent);
+        }
+
+        /// <summary>
+        /// Uploads a DNN package for local upgrade.
+        /// </summary>
+        /// <returns>Either a package info or an error message.</returns>
+        [HttpPost]
+        [IFrameSupportedValidateAntiForgeryToken]
+        public Task<HttpResponseMessage> Upload()
+        {
+            try
+            {
+                return this.UploadFileAction();
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex);
+                return Task.FromResult(this.Request.CreateErrorResponse(HttpStatusCode.InternalServerError, ex.Message));
+            }
+        }
+
+        private async Task<HttpResponseMessage> UploadFileAction()
```

Comment:
It would be good to take in a `CancellationToken` from the web API method and pass it through here, so that if the request gets cancelled this method can respond immediately.